bee-network/src/network/host.rs

```diff
@@ -0,0 +1,252 @@
+                internal_event_sender,
+                mut internal_command_receiver,
+                peerlist,
+            } = config;
+
+            let mut swarm = start_swarm(local_keys.clone(), bind_multiaddr, internal_event_sender.clone()).await;
+
+            tokio::spawn(async move {
+                loop {
+                    let swarm_next_event = Swarm::next_event(&mut swarm);
+                    let recv_command = (&mut internal_command_receiver).recv();
+
+                    tokio::select! {
+                        _ = &mut shutdown => break,
+                        event = swarm_next_event => {
+                            process_swarm_event(event, &internal_event_sender, &peerlist).await;
+                        }
+                        command = recv_command => {
+                            if let Some(command) = command {
+                                process_command(command, &mut swarm, &peerlist).await;
+                            }
+                        },
+                    }
+                }
+
+                info!("Network Host stopped.");
+            });
+
+            info!("Network Host started.");
+        }
+    }
+}
+
+async fn start_swarm(
+    local_keys: Keypair,
+    bind_multiaddr: Multiaddr,
+    internal_event_sender: InternalEventSender,
+) -> Swarm<SwarmBehavior> {
+    let mut swarm = build_swarm(&local_keys, internal_event_sender)
+        .await
+        .expect("Fatal error: creating transport layer failed.");
+
+    info!("Binding to: {}", bind_multiaddr);
+
+    let _ = Swarm::listen_on(&mut swarm, bind_multiaddr).expect("Fatal error: binding address failed.");
+
+    swarm
+}
+
+async fn process_swarm_event(
+    event: SwarmEvent<(), impl std::error::Error>,
+    internal_event_sender: &InternalEventSender,
+    peerlist: &PeerList,
+) {
+    match event {
+        SwarmEvent::NewListenAddr(multiaddr) => {
+            debug!("Swarm event: new listen address {}.", multiaddr);
+
+            internal_event_sender
+                .send(InternalEvent::AddressBound {
+                    address: multiaddr.clone(),
+                })
+                .expect("send error");
+
+            // INVESTIGATE
+            peerlist
+                .0
+                .write()
+                .await
+                .insert_local_addr(multiaddr)
+                .expect("insert_local_addr");
+        }
+        SwarmEvent::ConnectionEstablished { peer_id, .. } => {
+            debug!("Swarm event: connection established with {}.", alias!(peer_id));
+        }
+        SwarmEvent::ConnectionClosed { peer_id, .. } => {
+            debug!("Swarm event: connection closed with {}.", alias!(peer_id));
+        }
+        SwarmEvent::ListenerError { error } => {
+            error!("Swarm event: listener error {}.", error);
+        }
+        SwarmEvent::Dialing(peer_id) => {
+            // NB: strange, but this event is not actually fired when dialing. (open issue?)
```

Comment:
GH issue to not forget about it

Reply:
yea, no priority :smiley:  but good point. If I find the time to do that, then I am probably REALLY bored. haha